src/codegen.cpp

```diff
@@ -515,6 +516,50 @@ static inline void add_named_global(JuliaVariable *name, void *addr)
     add_named_global(name->name, addr);
 }
 
+static AttributeList get_attrs_ipoeffects(LLVMContext &context, std::pair<bool,uint32_t> effect, bool is_decl, bool has_ptr_arg)
+{
+    if (!effect.first)
+        return AttributeList();
+    uint32_t effects = effect.second;
+    uint8_t consistent = (effects >> 0) & 0x07;
+    uint8_t effect_free = (effects >> 3) & 0x03;
+    bool nothrow = (effects >> 5) & 0x01;
+    bool terminates = (effects >> 6) & 0x01;
+    bool notaskstate = (effects >> 7) & 0x01;
+    uint8_t inaccessiblememonly = (effects >> 8) & 0x03;
+    bool nonoverlayed = (effects >> 10) & 0x01;
+    bool noinbounds = (effects >> 11) & 0x01;
+    AttrBuilder attr(context);
+
+    if (consistent == 0) {
+    }
+    if (effect_free == 0) {
+        // if (!has_ptr_arg)
+            // attr.addAttribute(Attribute::ReadOnly); //Not legal because of the ptls
+    }
+    if (nothrow == 1) {
+        attr.addAttribute(Attribute::NoUnwind);
+    }
+    if (terminates == 1){
+        attr.addAttribute(Attribute::WillReturn);
+    }
+    if (inaccessiblememonly == 0) {
+        attr.addAttribute(Attribute::InaccessibleMemOrArgMemOnly); // Can't be inaccesiblememonly because of the ptls
+    } else if (inaccessiblememonly == 2){
+        attr.addAttribute(Attribute::InaccessibleMemOrArgMemOnly);
+    }
+    if (notaskstate == 1) {
+    }
+    if (nonoverlayed == 1) {
+    }
+    if (noinbounds == 1) {
+    }
+    if (is_decl && consistent == 0 && effect_free == 0 && nothrow == 1 && terminates == 1 && has_ptr_arg == false){
+        attr.addAttribute(Attribute::Speculatable); // This might not be legal because it assumes pointers can be null
+    }
+    attr.addAttribute(std::to_string(effects));
```

Comment:
This seems like it should be ifndef NDEBUG?

Reply:
Maybe we can add it as a comment to the module instead.